crates/e2e/src/builders.rs

```diff
@@ -0,0 +1,176 @@
+{
+    fn from(inner: CreateBuilderPartial<E, Args, R>) -> Self {
+        Self { inner }
+    }
+}
+
+impl<E: Environment, Args: Encode, R> ConstructorBuilder<E, Args, R> {
+    /// Returns encoded constructor arguments.
+    pub fn exec_input(self) -> Vec<u8> {
+        // set all the other properties to default values, we only require the `exec_input`.
+        self.inner
+            .endowment(0u32.into())
+            .code_hash(ink_primitives::Clear::clear())
+            .salt_bytes(Vec::new())
+            .params()
+            .exec_input()
+            .encode()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Message<E: Environment, RetType> {
+    account_id: E::AccountId,
+    exec_input: Vec<u8>,
+    marker: std::marker::PhantomData<RetType>,
+}
+
+impl<E, RetType> Message<E, RetType>
+where
+    E: Environment,
+{
+    pub fn build<F, Ref, Args>(
+        account_id: E::AccountId,
+        message: F,
+    ) -> Message<E, RetType>
+    where
+        E: Environment,
+        Ref: TraitCallBuilder + FromAccountId<E>,
+        Args: scale::Encode,
+        F: FnOnce(
+            &<Ref as TraitCallBuilder>::Builder,
+        ) -> CallBuilder<
+            E,
+            Set<Call<E>>,
+            Set<ExecutionInput<Args>>,
+            Set<ReturnType<RetType>>,
+        >,
+        RetType: scale::Decode,
+    {
+        let contract_ref = <Ref as FromAccountId<E>>::from_account_id(account_id.clone());
+        let call_builder = <Ref as TraitCallBuilder>::call(&contract_ref);
+        let builder = message(&call_builder);
+        let exec_input = builder.params().exec_input().encode();
+        Message {
+            account_id,
+            exec_input,
+            marker: Default::default(),
+        }
+    }
+
+    pub fn account_id(&self) -> &E::AccountId {
+        &self.account_id
+    }
+
+    pub fn exec_input(&self) -> &[u8] {
+        &self.exec_input
+    }
+}
+
+/// Convenience method for building messages for the default environment
+pub fn build_message<Ref>(
+    account_id: <ink_env::DefaultEnvironment as Environment>::AccountId,
+) -> MessageBuilder<ink_env::DefaultEnvironment, Ref>
+where
+    Ref: TraitCallBuilder + FromAccountId<ink_env::DefaultEnvironment>,
+{
+    MessageBuilder::from_account_id(account_id)
+}
+
+/// Build messages using a contract ref.
+pub struct MessageBuilder<E: Environment, Ref> {
+    account_id: E::AccountId,
+    contract_ref: Ref,
+}
+
+impl<E, Ref> MessageBuilder<E, Ref>
+where
+    E: Environment,
+    Ref: TraitCallBuilder + FromAccountId<E>,
+{
+    pub fn from_account_id(account_id: E::AccountId) -> Self {
+        let contract_ref = <Ref as FromAccountId<E>>::from_account_id(account_id.clone());
+        Self {
+            account_id,
+            contract_ref,
+        }
+    }
+
+    pub fn account_id(&self) -> &E::AccountId {
+        &self.account_id
+    }
+
+    pub fn call<F, Args, RetType>(mut self, mut message: F) -> Message<E, RetType>
```

Comment:
Would be nice to even have a doc comment showing the expected usage pattern

Reply:
@ascjones FYI, I made this compile

Reply:
:see_no_evil: